File: 4-Behavioral Patterns/7-Mediator Pattern/Phone Exchange Example (1-to-1)/Java/runner/Avengers.java
package runner;

import model.MemberPhoneCustomer;
import model.NormalPhoneCustomer;
import model.PhoneCustomer;
import model.PhoneExchange;

/**
 * Application that actually uses Mediator Pattern.
 *
 * @author Ziang Lu
 */
public class Avengers {

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        PhoneExchange phoneExchange = new PhoneExchange();
        PhoneCustomer steve = new NormalPhoneCustomer("Steve", phoneExchange);
        System.out.println(steve);
        PhoneCustomer tony = new MemberPhoneCustomer("Tony", phoneExchange);
        System.out.println(tony);
        PhoneCustomer bruce = new NormalPhoneCustomer("Bruce", phoneExchange);
        System.out.println(bruce);

        System.out.println();

        String stevePhoneNumber = steve.getNumber();
        tony.call(stevePhoneNumber, "Come back.");
        System.out.println();
        bruce.call(stevePhoneNumber, "Tony needs you.");

        /*
         * Output:
         * NormalPhoneCustomer{name=Steve, phoneNumber=855-665-7169} with 50 calls per month
         * MemberPhoneCustomer{name=Tony, phoneNumber=434-316-8721} with unlimited calls per month
         * NormalPhoneCustomer{name=Bruce, phoneNumber=439-606-6158} with 50 calls per month
         *
         * Tony calling 855-665-7169 with message: Come back. ...
         * Successful!
         * Steve received a call from 434-316-8721 with message: Come back.
         *
         * Bruce calling 855-665-7169 with message: Tony needs you. ...
         * Successful!
         * Steve received a call from 439-606-6158 with message: Tony needs you.
         */
    }

}

File: 4-Behavioral Patterns/9-Mediator Pattern/Phone Exchange Example (1-to-1)/Python/avengers.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Application that actually uses Mediator Pattern.
"""

__author__ = 'Ziang Lu'

from model import MemberPhoneCustomer, NormalPhoneCustomer, PhoneExchange


def main():
    phone_exchange = PhoneExchange()
    steve = NormalPhoneCustomer(name='Steve', phone_exchange=phone_exchange)
    print(steve)
    tony = MemberPhoneCustomer(name='Tony', phone_exchange=phone_exchange)
    print(tony)
    bruce = NormalPhoneCustomer(name='Bruce', phone_exchange=phone_exchange)
    print(bruce)

    print()

    steve_phone_number = steve.phone_number
    tony.call(recipient_number=steve_phone_number, msg='Come back.')
    print()
    bruce.call(recipient_number=steve_phone_number, msg='Tony needs you.')


if __name__ == '__main__':
    main()

# Output:
# NormalPhoneCustomer[name=Steve, phone_number=381-622-2946 with 50 calls per month]
# MemberPhoneCustomer[name=Tony, phone_number=960-265-8658 with unlimited calls per month
# NormalPhoneCustomer[name=Bruce, phone_number=670-355-4820 with 50 calls per month]
#
# Tony calling 381-622-2946 with the message: Come back. ...
# Successful!
# Steve received a call from 960-265-8658 with message: Come back.
#
# Bruce calling 381-622-2946 with the message: Tony needs you. ...
# Successful!
# Steve received a call from 670-355-4820 with message: Tony needs you.
